.github/workflows/build_and_test.yml

```diff
@@ -0,0 +1,137 @@
+                  echo "Secret created successfully"
+              else
+                  echo "Failed to create secret"
+                  exit 1
+              fi
+          fi
+
+      - name: checkout
+        uses: actions/checkout@v5
+
+      - name: Deploy CodePipeline to AWS CloudFormation
+        uses: aws-actions/aws-cloudformation-github-deploy@v1
+        id: pipeline-deploy
+        with:
+          template: configuration/build/codepipeline_cf_template.yaml
+          name: Synapse-Build-${{ env.user }}
+          capabilities: CAPABILITY_IAM,CAPABILITY_NAMED_IAM
+          parameter-overrides: >-
+            SynapseBranch=${{ env.branch }},
+            SynapseRepoOwner=${{ github.repository_owner }},
+            GitHubToken=${{ secrets.GITHUB_TOKEN }},
+            BuildParamUser=${{ env.user }},
+            StackBuilderBranch=${{ env.stack_builder_branch }},
+            StackBuilderRepoOwner=${{ env.stack_builder_repo_owner }},
+            FullBuild=${{ env.full_build }}
+
+      - name: Trigger CodePipeline
+        run: |
+          EXECUTION_ID=$(aws codepipeline start-pipeline-execution \
+            --name ${{ steps.pipeline-deploy.outputs.PipelineName }} \
+            --source-revisions actionName=CheckoutSynapseRepo,revisionType=COMMIT_ID,revisionValue=${{ github.sha }} \
+            --query 'pipelineExecutionId' --output text)
+
+          echo "Pipeline execution ID: $EXECUTION_ID"
+
+          # Wait for completion
+          FINAL_STAGE="BuildSynapseRepo"
+          while true; do
+            # this is a list of the statuses of all the stages that have started
+            ALL_STAGE_STATUSES=$(aws codepipeline get-pipeline-state --name ${{ steps.pipeline-deploy.outputs.PipelineName }} \
+              --query "stageStates[?(latestExecution.pipelineExecutionId=='$EXECUTION_ID')].latestExecution.status" --output text)
+            # this is the status of just the final stage, or an empty sting if it hasn't started
+            FINAL_STAGE_STATUS=$(aws codepipeline get-pipeline-state --name ${{ steps.pipeline-deploy.outputs.PipelineName }} \
+              --query "stageStates[?(latestExecution.pipelineExecutionId=='$EXECUTION_ID' && stageName=='$FINAL_STAGE')].latestExecution.status" --output text)
+
+            if [[ "$ALL_STAGE_STATUSES" =~ "Failed" || "$ALL_STAGE_STATUSES" =~ "Stopped" ]]; then
+              # if there is a failure in any stage we stop
+              echo "Pipeline execution $EXECUTION_ID finished with stage statuses: $ALL_STAGE_STATUSES"
+              exit 1
+            elif [[ "$FINAL_STAGE_STATUS" == "InProgress" || "$FINAL_STAGE_STATUS" == "" ]]; then
+              # if the final stage hasn't started yet or is in progress, we continue
+              echo "Pipeline execution $EXECUTION_ID is still in progress. Current status: $STATUS. Waiting..."
+              sleep 30 # Wait for 30 seconds before checking again
+            elif [[ "$FINAL_STAGE_STATUS" == "Succeeded" ]]; then
```

Comment:
Should  we move Succeeded check before checking InProgress because that waits for 30 seconds